### ISSUE TITLE
At startup `ERROR: editor/progress_dialog.cpp:169 - Do not use progress dialog (task) while flushing the message queue or using call_deferred()!`

### ISSUE THREAD

**Issue by miv391** [godotengine/godot#100900]
### Tested versions

- Happens in v4.4.dev7.mono.official [46c8f8c5c]

### System information

Godot v4.4.dev7.mono - Windows 10 (build 19045) - Multi-window, 1 monitor - Vulkan (Forward+) - dedicated NVIDIA GeForce GTX 1060 6GB (NVIDIA; 32.0.15.6094) - Intel(R) Core(TM) i5-3570K CPU @ 3.40GHz (4 threads)

### Issue description

I get this error sometimes at startup. My project is mostly GDScript code with two C# scripts.

This _might_ be related to https://github.com/godotengine/godot/issues/96550, at least the error message is the same.

```
Godot Engine v4.4.dev7.mono.official (c) 2007-present Juan Linietsky, Ariel Manzur & Godot Contributors.
--- Debug adapter server started on port 6006 ---
--- GDScript language server started on port 6005 ---
  ERROR: editor/progress_dialog.cpp:169 - Do not use progress dialog (task) while flushing the message queue or using call_deferred()!
  ERROR: editor/progress_dialog.cpp:203 - Condition "!tasks.has(p_task)" is true. Returning: canceled
  ERROR: editor/progress_dialog.cpp:203 - Condition "!tasks.has(p_task)" is true. Returning: canceled
  ERROR: editor/progress_dialog.cpp:203 - Condition "!tasks.has(p_task)" is true. Returning: canceled
  ERROR: editor/progress_dialog.cpp:203 - Condition "!tasks.has(p_task)" is true. Returning: canceled
  ERROR: editor/progress_dialog.cpp:203 - Condition "!tasks.has(p_task)" is true. Returning: canceled
  ERROR: editor/progress_dialog.cpp:203 - Condition "!tasks.has(p_task)" is true. Returning: canceled
  ERROR: editor/progress_dialog.cpp:226 - Condition "!tasks.has(p_task)" is true.
[LSP] Connection Taken
```

### Steps to reproduce

Unfortunately I have no idea. I _feel_ that I have mostly seen this error when starting Godot first time after booting the computer. I did managed to get the error once by repeatedly starting Godot while Microsoft Defender was scanning the drive thus making everything slow. But even then I got the error only once with maybe 15 tries.

### Minimal reproduction project (MRP)

N/A

**Comment by MikeSchulze** [comment]
It occurs just on an empty project to enforce rebuilding the cache
```
$ $GODOT_BIN --path . -e --headless --quit-after 200
Godot Engine v4.4.dev7.mono.official.46c8f8c5c - https://godotengine.org

ERROR: Do not use progress dialog (task) while flushing the message queue or using call_deferred()!
   at: add_task (editor/progress_dialog.cpp:169)
ERROR: Condition "!tasks.has(p_task)" is true. Returning: canceled
   at: task_step (editor/progress_dialog.cpp:203)
ERROR: Condition "!tasks.has(p_task)" is true. Returning: canceled
   at: task_step (editor/progress_dialog.cpp:203)
ERROR: Condition "!tasks.has(p_task)" is true. Returning: canceled
   at: task_step (editor/progress_dialog.cpp:203)
ERROR: Condition "!tasks.has(p_task)" is true. Returning: canceled
   at: task_step (editor/progress_dialog.cpp:203)
ERROR: Condition "!tasks.has(p_task)" is true. Returning: canceled
   at: task_step (editor/progress_dialog.cpp:203)
ERROR: Condition "!tasks.has(p_task)" is true. Returning: canceled
   at: task_step (editor/progress_dialog.cpp:203)
ERROR: Condition "!tasks.has(p_task)" is true.
   at: end_task (editor/progress_dialog.cpp:226)

```

**Comment by MadeScientist** [comment]
`Godot Engine v4.4.rc1.official (c) 2007-present Juan Linietsky, Ariel Manzur & Godot Contributors.
--- Debug adapter server started on port 6006 ---
--- GDScript language server started on port 6005 ---
  ERROR: editor/progress_dialog.cpp:183 - Do not use progress dialog (task) while flushing the message queue or using call_deferred()!
  ERROR: editor/progress_dialog.cpp:217 - Condition "!tasks.has(p_task)" is true. Returning: canceled
  ERROR: editor/progress_dialog.cpp:217 - Condition "!tasks.has(p_task)" is true. Returning: canceled
  ERROR: editor/progress_dialog.cpp:217 - Condition "!tasks.has(p_task)" is true. Returning: canceled
  ERROR: editor/progress_dialog.cpp:217 - Condition "!tasks.has(p_task)" is true. Returning: canceled
  ERROR: editor/progress_dialog.cpp:217 - Condition "!tasks.has(p_task)" is true. Returning: canceled
  ERROR: editor/progress_dialog.cpp:217 - Condition "!tasks.has(p_task)" is true. Returning: canceled
  ERROR: editor/progress_dialog.cpp:240 - Condition "!tasks.has(p_task)" is true.
`

**Comment by pewcworrell** [comment]
For us I narrowed it down to a plugin that was using call_deferred. We changed
```
editor.get_resource_filesystem().scan.call_deferred()
```
to
```
editor.get_resource_filesystem().scan
```
and these errors went away.

**Comment by pewcworrell** [comment]
I ended up running into this issue anyways while exporting with --headless. I built a script so you always get the error. To run, set an environment variable for your Godot console executable path to be GODOT_EXEC. Then open up the builds folder and run the script, and observe the errors in the output. [test.zip](https://github.com/user-attachments/files/18972824/test.zip). 

Given how I got the error beforehand in a plugin that had call_deferred, I think @Riteo is spot on.

I'm not sure if the solution would be to change the --headless to not make any deferred calls, or to make the deferred calls not be an issue.

**Comment by mendsley** [comment]
I _think_ this originates in `NOTIFICATION_ENTER_TREE` of `EditorNode::_notification`
https://github.com/godotengine/godot/blob/01545c995b0612c68f9dfce8f6cc67576c298381/editor/editor_node.cpp#L687

This eventually causes the error in the callstack (`_begin_first_scan` was tail call optimized out of the stack into `scan`):
```
>	godot.windows.editor.x86_64.exe!ProgressDialog::add_task(const String & p_task, const String & p_label, int p_steps, bool p_can_cancel) Line 183	C++
 	godot.windows.editor.x86_64.exe!EditorNode::progress_add_task(const String & p_task, const String & p_label, int p_steps, bool p_can_cancel) Line 4950	C++
 	godot.windows.editor.x86_64.exe!EditorProgress::EditorProgress(const String & p_task, const String & p_label, int p_amount, bool p_can_cancel, bool p_force_background) Line 192	C++
 	godot.windows.editor.x86_64.exe!EditorFileSystem::_first_scan_filesystem() Line 301	C++
 	godot.windows.editor.x86_64.exe!EditorFileSystem::scan() Line 1066	C++
 	[Inline Frame] godot.windows.editor.x86_64.exe!call_with_variant_args_helper(EditorNode *) Line 315	C++
 	[Inline Frame] godot.windows.editor.x86_64.exe!call_with_variant_args(EditorNode * p_instance, void(EditorNode::*)() p_method, const Variant * *) Line 429	C++
 	godot.windows.editor.x86_64.exe!CallableCustomMethodPointer<EditorNode,void>::call(const Variant * * p_arguments, int p_argcount, Variant & r_return_value, Callable::CallError & r_call_error) Line 109	C++
 	godot.windows.editor.x86_64.exe!Callable::callp(const Variant * * p_arguments, int p_argcount, Variant & r_return_value, Callable::CallError & r_call_error) Line 58	C++
 	godot.windows.editor.x86_64.exe!CallQueue::_call_function(const Callable & p_callable, const Variant * p_args, int p_argcount, bool p_show_error) Line 221	C++
 	godot.windows.editor.x86_64.exe!CallQueue::flush() Line 284	C++
 	godot.windows.editor.x86_64.exe!SceneTree::physics_process(double p_time) Line 536	C++
 	godot.windows.editor.x86_64.exe!Main::iteration() Line 4443	C++
 	godot.windows.editor.x86_64.exe!OS_Windows::run() Line 2062	C++
 	godot.windows.editor.x86_64.exe!widechar_main(int argc, wchar_t * * argv) Line 97	C++
 	godot.windows.editor.x86_64.exe!_main() Line 124	C++
 	godot.windows.editor.x86_64.exe!main(int argc, char * * argv) Line 136	C++
```

**Comment by mendsley** [comment]
Caveat: While this does workaround the issue above, I'm not totally satisfied with it as a solution.

Most of this appears to stem from the ProgressDialog. It looks like everything using EditorNode::progress_dialog is wrapped in a null check, so for the time I simply suppress the creation of the ProgressDialog in headless runs.

It's feels like a timing/race issue with the deferred calls in ProgressDialog. It's deferring tasks which make sense, but headless mode can end up flushing that queue much sooner than would be the case in the normal Editor flow.

Disabling the ProgressDialog:
https://github.com/mendsley/godot/commit/bb575543e0e81dafcd33892490e3d5046dbfcf2b

**Comment by mihe** [comment]
I made a formal issue (#103398) for the case when doing a headless import, as mentioned [above](https://github.com/godotengine/godot/issues/100900#issuecomment-2567712127), since I figured it might be different from the issue reported here.

I also did some bisecting and found that the headless import errors is a regression caused by #93064.

**Comment by lnd1** [comment]
(Please note that I'm very new to the developer space and I know pretty much nothing about how Godot works, or coding in general.)
**Bug description:**
 I get several errors if I try to minimize a game/project while it is loading. Even if I create a new project. I'm guessing it has something to do with the above bugs since they have some of the same error messages.
**Test info:**
 All tests were done on a PC with the windows OS and windows Godot versions. Unless stated otherwise, all tests were done on Forward+ renderer mode. (Last updated this 7/1/25 and it still occurs in the latest versions and snapshots)
**Versions tested which this occurs in:**
v4.5.beta1.official [46c495ca2]
v4.5.dev5.official [64b09905c]
v4.5.dev4.official [209a446e3]
v4.5.dev3.official [55fc6e73f]
v4.5.dev2.official [af2c71397]
v4.5.dev1.official [97241ffea]
v4.4.1.stable.official [49a5bc7b6]
v4.4.stable.official [4c311cbee]: Forward+, Mobile and Compatibility modes
v4.4.rc3.official [15ff45068]
v4.4.rc2.official [01545c995]
v4.4.rc1.official [8ed125b42]
v4.4.dev7.official [46c8f8c5c]
v4.4.dev6.official [1f47e4c4e]
v4.4.dev5.official [9e6098432]
v4.4.dev4.official [36e6207bb]
v4.4.dev3.official [f4af8201b]
v4.4.dev2.official [97ef3c837]
**Versions tested which this does _Not_ occur in:**
v4.4.dev1.official [28a72fa43]
4.3 Stable
4.2.2 Stable

Also I'll just ask: Should I move this comment into a proper bug report or leave it as a comment?

Here is a video of me getting these errors in 4.4 Stable:

https://github.com/user-attachments/assets/fe87cfff-e4c8-4261-a8b9-bb2cacea4f44

**Comment by CreatedBySeb** [comment]
Not getting this issue anymore on `v4.5.dev.mono.custom_build.215acd52e`, looks like #103403 may have fixed it but not been marked as such

```
Godot v4.5.dev.mono (215acd52e) - Pop!_OS 22.04 LTS on X11 - X11 display driver, Multi-window, 2 monitors - OpenGL 3 (Compatibility) - NVIDIA GeForce RTX 3060 (nvidia; 570.133.07) - AMD Ryzen 7 3700X 8-Core Processor (16 threads)
```

**Comment by AThousandShips** [comment]
That PR didn't fix this, as per [this comment](https://github.com/godotengine/godot/pull/103403#issuecomment-2706366202), so this might not be completely fixed

CC @mihe

**Comment by mihe** [comment]
It was indeed not fixed by #103403, since it only really dealt with `--headless`. Last I checked the repro provided by @lnd1 still reliably reproduces this issue.

**Comment by KoBeWi** [comment]
See if #108893 fixes the issue for you.

**Comment by lnd1** [comment]
> See if [#108893](https://github.com/godotengine/godot/pull/108893) fixes the issue for you.

Sorry for my ignorance, but could you tell me how to test the changes?

**Comment by KoBeWi** [comment]
If you don't have a setup for making custom builds, you can grab an executable from here: https://github.com/godotengine/godot/actions/runs/16509107737?pr=108893 (see bottom of the page)

**Comment by lnd1** [comment]
> If you don't have a setup for making custom builds, you can grab an executable from here: https://github.com/godotengine/godot/actions/runs/16509107737?pr=108893 (see bottom of the page)

I just download it and run it right? The editor .exe is crashing if I try to open or make a project.

**Comment by deralmas** [comment]
@lnd1 Does it say where it crashes? IIRC those builds are mono enabled and might need mono to functions but I'm not entirely sure about that.

**Comment by lnd1** [comment]
> [@lnd1](https://github.com/lnd1) Does it say where it crashes? IIRC those builds are mono enabled and might need mono to functions but I'm not entirely sure about that.

I don't think so? Here is a video of what I did:

https://github.com/user-attachments/assets/af02c073-54c6-4925-b030-a4ade8ba0fcd

**Comment by KoBeWi** [comment]
I can confirm the crash. Probably something wrong with the build, it spams shader errors in the console.

**Comment by WhalesState** [comment]
I can confirm the crash too with that build from GH actions, but it doesn't happen here https://github.com/godotengine/godot/actions/runs/16478237489 which has the same fix before rebasing. tested with `scons dev_build=yes` before applying the comment changes and after to make sure it's not related to https://github.com/godotengine/godot/pull/108893, and will confirm if it also happens with local build from master.

**Comment by WhalesState** [comment]
Note: The link is not from a reviewed or merged PR so please don't use it for testing.

I have tested with this one because there was no new commits merged after the fix and it doesn't crash: https://github.com/godotengine/godot/actions/runs/16536329432
